state/linklayerdevices.go

```diff
@@ -191,6 +191,35 @@ func (dev *LinkLayerDevice) parentDocID() string {
 	return dev.st.docID(parentGlobalKey)
 }
 
+// SetProviderIDOps returns the operations required to set the input
+// provider ID for the link-layer device.
+func (dev *LinkLayerDevice) SetProviderIDOps(id network.Id) ([]txn.Op, error) {
+	// We only set the provider ID if it was previously empty.
+	if dev.doc.ProviderID != "" || id == "" || dev.doc.ProviderID == id.String() {
+		return nil, nil
+	}
+
+	// Since we assume that we are now setting the ID for the first time,
+	// ensure that it has not already been used to identify another device.
+	exists, err := dev.st.networkEntityGlobalKeyExists("linklayerdevice", id)
```

Comment:
This method makes a DB query which runs outside the generated txn ops below. Can we make it return a txn.Op instead to completely eliminate the possibility of a dup ID entry?

Reply:
`networkEntityGlobalKeyOp` used to insert the ID below already asserts that the doc is missing.

Remember that this method is being called from with the model operation `Build` command, so we will error at this point with an appropriate message if a dup is dedected, but the assertion still handles a race.